### PR TITLE
Testing Suite for JS library

### PR DESCRIPTION
**Pull request by ShubhamPalriwala**
## What does this PR do?
Adds Unit Testing to the `@formbricks/js` npm package with Jest and Babel!

Update (19th June): 
- [x] The tests look flaky due to varying response time hence that is yet to be fixed!

Update (21st June):
The tests have been migrated to use a mock API so are not flaky at all (tested by concurrently running 21 times)

Tests the following cases:
```
 PASS  tests/index.test.ts (9.381 s)
  ✓ Test Jest (4 ms)
  ✓ Formbricks should Initialise (2062 ms)
  ✓ Formbricks should get no current person
  ✓ Formbricks should set attributes (1893 ms)
  ✓ Formbricks should update attribute (572 ms)
  ✓ Formbricks should track event (1014 ms)
  ✓ Formbricks should refresh (624 ms)
  ✓ Formbricks should register for route change (5 ms)
  ✓ Formbricks should logout (812 ms)
```

Coverage we attain as of now:
```
-----------------------------------|---------|----------|---------|---------|-------------------------------------------------------------
File                               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-----------------------------------|---------|----------|---------|---------|-------------------------------------------------------------
All files                          |   42.87 |     18.8 |   33.15 |   44.16 |
 src                               |   82.85 |      100 |   72.72 |   82.85 |
  App.tsx                          |       0 |      100 |       0 |       0 | 16-25
  index.ts                         |     100 |      100 |     100 |     100 |
 src/components                    |     4.8 |        0 |       0 |    5.02 |
  CTAQuestion.tsx                  |       0 |        0 |       0 |       0 | 15-52
  FormbricksSignature.tsx          |       0 |      100 |       0 |       0 | 4
  Headline.tsx                     |       0 |      100 |       0 |       0 | 12
  HtmlBody.tsx                     |       0 |      100 |       0 |       0 | 5
  Modal.tsx                        |       0 |        0 |       0 |       0 | 14-22
  MultipleChoiceMultiQuestion.tsx  |       0 |        0 |       0 |       0 | 22-114
  MultipleChoiceSingleQuestion.tsx |       0 |        0 |       0 |       0 | 22-69
  NPSQuestion.tsx                  |       0 |        0 |       0 |       0 | 17-58
  OpenTextQuestion.tsx             |       0 |      100 |       0 |       0 | 20-28
  Progress.tsx                     |       0 |      100 |       0 |       0 | 4
  QuestionConditional.tsx          |       0 |        0 |       0 |       0 | 23
  RatingQuestion.tsx               |       0 |        0 |       0 |       0 | 34-191
  Smileys.tsx                      |   52.38 |      100 |       0 |   52.38 | 5,66,127,197,227,256,289,321,367,419
  Subheader.tsx                    |       0 |      100 |       0 |       0 | 4
  SubmitButton.tsx                 |       0 |        0 |       0 |       0 | 7
  SurveyView.tsx                   |       0 |        0 |       0 |       0 | 24-232
  ThankYouCard.tsx                 |       0 |      100 |       0 |       0 | 12
 src/lib                           |    55.9 |       40 |   54.25 |   57.94 |
  api.ts                           |   16.66 |        0 |       0 |   16.66 | 5-12
  cleanHtml.ts                     |       0 |        0 |       0 |       0 | 13-78
  commandQueue.ts                  |   89.65 |     64.7 |     100 |     100 | 15,30,45-55
  config.ts                        |   93.33 |     62.5 |     100 |   92.85 | 33
  display.ts                       |   13.33 |        0 |       0 |   13.33 | 8-30,37-55
  errors.ts                        |   71.87 |       50 |   64.28 |      75 | 18-19,30,55,105-106,131
  event.ts                         |   70.58 |       50 |     100 |   70.58 | 30-32,51-53
  init.ts                          |   63.07 |    48.38 |      60 |   65.07 | 28,32,50-51,58-60,76-98,109,148
  logger.ts                        |   86.66 |    83.33 |   83.33 |   86.66 | 34,45
  noCodeEvents.ts                  |   46.47 |     25.8 |   31.25 |   48.43 | 18,26,34-36,65-80,87,94-125
  person.ts                        |      76 |    59.37 |      80 |   77.77 | 35,51,71,88,110,125,133,144-145,148,171-172,186,191,210,215
  response.ts                      |   14.28 |        0 |       0 |   14.28 | 9-29,37-57
  session.ts                       |   25.71 |        0 |      25 |   28.12 | 13-41,51-52,56-74
  settings.ts                      |   82.35 |       50 |     100 |    87.5 | 20-22
  styles.ts                        |     100 |       50 |     100 |     100 | 6
  utils.ts                         |    9.09 |        0 |       0 |    9.09 | 2-16
  widget.ts                        |    42.3 |        0 |   16.66 |    42.3 | 16-27,36-48
 tests                             |     100 |      100 |     100 |     100 |
  constants.ts                     |     100 |      100 |     100 |     100 |
-----------------------------------|---------|----------|---------|---------|-------------------------------------------------------------
```

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?
- Update the `packages/js/tests/constants.ts` file with your `environmentId`
- Run `npm run test` inside the `packages/js` directory

## Checklist
- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by ShubhamPalriwala** [comment]
@mattinannt, I added a small if conditional check in the library while I was dealing with responses and mocking them! Please take a look and let me know if that can have adverse effects!

**Review by mattinannt — COMMENTED** [review]
Thanks for the changes 😊 Fetch-Mock works great and it's so cool that we now have tests for the formbricks-js package :-) feels great!
I made a few comments and I think the `Formbricks should set attributes` case could be split into multiple cases that we can test independently, but I think it works for now 😊.
Can you please also add the `pnpm test` command to turborepo so we can run `pnpm test` in the project root and it will run all tests from all apps/packages (if there are any)?

I'm also not so happy about all the console.logs from the widget in the console during tests. I know we currently need them for the checks in the tests, but can we maybe prevent the logging, e.g. by mocking the `logger.debug()` function or something like that?

File: packages/js/tests/__mocks__/apiMock.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+export const mockInitResponse = () => {
+    fetchMock.mockResponseOnce(JSON.stringify({
+        apiHost,
+        environmentId,
+        person: {
+            id: initialPersonUid,
+            environmentId,
+            attributes: []
+        },
+        session: {},
```
[the remark this answers is not in the document]
Comment (ShubhamPalriwala) [reply packages/js/tests/__mocks__/apiMock.ts]: resolved in e452a15

File: packages/js/tests/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    expect(configFromBrowser).toBeTruthy();
+
+    if (configFromBrowser) {
+        const jsonSavedConfig = JSON.parse(configFromBrowser);
+        expect(jsonSavedConfig.environmentId).toStrictEqual(environmentId);
+        expect(jsonSavedConfig.apiHost).toStrictEqual(apiHost);
+    }
+});
+
+test("Formbricks should get no current person", () => {
```
[the remark this answers is not in the document]
Comment (ShubhamPalriwala) [reply packages/js/tests/index.test.ts]: resolved in 3089508

**Comment by ShubhamPalriwala** [comment]
Hey @mattinannt, I have made all the changes and suggestions you gave! Only left with the integration of `cuid2` since I'm unable to set it up for the mock js dom environment we are creating. Can you help me with that please? 🙏🏼

**Review by mattinannt — APPROVED** [review]
🎉💪 Ready to merge! 😊

**Pull request merged**